### PR TITLE
TL/UCP: allgather ring

### PR DESCRIPTION
**Pull request by vspetrov**
## What
Adds ring allgather algorithm into TL/UCP + test_mpi::allgather
Also adds ucc_mc_memcpy interface which is used by the allgather and potentially will be used in other algs/tls/cls.

### CODE REVIEW

File: src/components/tl/ucp/allgather/allgather.c [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+#include "config.h"
+#include "tl_ucp.h"
+#include "allgather.h"
+
+ucc_status_t ucc_tl_ucp_allgather_ring_start(ucc_coll_task_t *task);
+ucc_status_t ucc_tl_ucp_allgather_ring_progress(ucc_coll_task_t *task);
+
+ucc_status_t ucc_tl_ucp_allgather_init(ucc_tl_ucp_task_t *task)
+{
+    task->super.post     = ucc_tl_ucp_allgather_ring_start;
```
Comment (Sergei-Lebedev) [inline src/components/tl/ucp/allgather/allgather.c:15]: check that datatype is not user defined

File: src/components/mc/cpu/mc_cpu.c [hunk 2]
```diff
@@ -70,6 +70,14 @@ static ucc_status_t ucc_mc_cpu_mem_free(void *ptr)
     return UCC_OK;
 }
 
+static ucc_status_t ucc_mc_cpu_mem_copy(void *dst, const void *src, size_t len,
+                                        ucc_memory_type_t dst_mem, /* NOLINT */
+                                        ucc_memory_type_t src_mem) /* NOLINT */
+{
+    memcpy(dst, src, len);
```
[the remark this answers is not in the document]
Comment (vspetrov) [reply src/components/mc/cpu/mc_cpu.c:77]: 👍

File: src/core/ucc_mc.c [hunk 3]
```diff
@@ -134,6 +134,25 @@ ucc_status_t ucc_mc_free(void *ptr, ucc_memory_type_t mem_type)
...
+                           ucc_memory_type_t dst_mem,
+                           ucc_memory_type_t src_mem)
+{
+    if (src_mem == UCC_MEMORY_TYPE_HOST &&
+         dst_mem == UCC_MEMORY_TYPE_HOST) {
+        UCC_CHECK_MC_AVAILABLE(UCC_MEMORY_TYPE_HOST);
+        return mc_ops[UCC_MEMORY_TYPE_HOST]->mem_copy(dst, src, len,
+                                                      UCC_MEMORY_TYPE_HOST,
+                                                      UCC_MEMORY_TYPE_HOST);
+    } else {
```
[the remark this answers is not in the document]
Comment (vspetrov) [reply src/core/ucc_mc.c:147]: 👍

File: src/core/ucc_mc.c [hunk 4]
```diff
@@ -134,6 +134,25 @@ ucc_status_t ucc_mc_free(void *ptr, ucc_memory_type_t mem_type)
     return mc_ops[mem_type]->mem_free(ptr);
 }
 
+ucc_status_t ucc_mc_memcpy(void *dst, const void *src, size_t len,
+                           ucc_memory_type_t dst_mem,
+                           ucc_memory_type_t src_mem)
+{
+    if (src_mem == UCC_MEMORY_TYPE_HOST &&
```
[the remark this answers is not in the document]
Comment (vspetrov) [reply src/core/ucc_mc.c:141]: 👍

File: src/components/tl/ucp/allgather/allgather_ring.c [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    void              *rbuf      = task->args.dst.info.buffer;
+    ucc_memory_type_t  smem      = task->args.src.info.mem_type;
+    ucc_memory_type_t  rmem      = task->args.dst.info.mem_type;
+    ucc_datatype_t     dt        = task->args.dst.info.datatype;
+    size_t             data_size = count * ucc_dt_size(dt);
+    ucc_status_t       status;
+
+    task->super.super.status     = UCC_INPROGRESS;
+    if (!UCC_IS_INPLACE(task->args)) {
+        ucc_mc_memcpy((void*)((ptrdiff_t)rbuf + data_size * team->rank),
```
[the remark this answers is not in the document]
Comment (vspetrov) [reply src/components/tl/ucp/allgather/allgather_ring.c:65]: 👍

File: src/components/mc/cuda/mc_cuda.c [hunk 6]
```diff
@@ -78,6 +78,66 @@ static ucc_status_t ucc_mc_cuda_mem_free(void *ptr)
...
+            cudaMemcpyHostToDevice : cudaMemcpyDeviceToDevice;
+        break;
+    default:
+        status = UCC_ERR_INVALID_PARAM;
+        break;
+    }
+    return status;
+}
+
+static ucc_status_t ucc_mc_cuda_mem_cpy(void *dst, const void *src, size_t len,
```
[the remark this answers is not in the document]
Comment (vspetrov) [reply src/components/mc/cuda/mc_cuda.c:104]: 👍

File: src/components/tl/ucp/allgather/allgather.c (hunk 1 above)
Comment (vspetrov) [reply src/components/tl/ucp/allgather/allgather.c:21]: 👍

**Review by Sergei-Lebedev — APPROVED** [review]

File: src/components/mc/cuda/mc_cuda.c [hunk 7]
```diff
@@ -78,6 +78,66 @@ static ucc_status_t ucc_mc_cuda_mem_free(void *ptr)
...
+                                       ucc_memory_type_t dst_mem,
+                                       ucc_memory_type_t src_mem)
+{
+    cudaError_t    st;
+    cudaMemcpyKind kind;
+    ucc_status_t   status;
+    ucc_assert(dst_mem == UCC_MEMORY_TYPE_CUDA ||
+               src_mem == UCC_MEMORY_TYPE_CUDA);
+
+    if (UCC_OK != (status = ucc_mc_cuda_memcpy_kind_map(dst_mem, src_mem, &kind))) {
```
Comment (bureddy) [inline src/components/mc/cuda/mc_cuda.c:114]: I think we can skip finding the kind here. we can use cudaMemCpyDefault instead. which is recommended (https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1g85073372f776b4c4d5f89f7124b7bf79)

File: test/mpi/test_allgather.cc [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+    args.src.info.buffer      = sbuf;
+    args.src.info.count       = count;
+    args.src.info.datatype    = TEST_DT;
+    args.src.info.mem_type    = _mt;
+
+    args.dst.info.buffer      = rbuf;
+    args.dst.info.count       = count;
+    args.dst.info.datatype    = TEST_DT;
+    args.dst.info.mem_type    = _mt;
```
Comment (bureddy) [inline test/mpi/test_allgather.cc:46]: spaces before =

File: src/components/mc/cuda/mc_cuda.c (hunk 7 above)
Comment (vspetrov) [reply src/components/mc/cuda/mc_cuda.c:114]: @bureddy does that imply additional ptr kind detection inside libcuda? If we already know the ptr types do we want to do the checks again? Maybe it is cheap though, i'm not sure - what is your thought? @Sergei-Lebedev ? Also :  "However, cudaMemcpyDefault is only allowed on systems that support unified virtual addressing" - is this a limitation or we can ignore that ?

File: src/components/tl/ucp/allgather/allgather_ring.c [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    ucc_rank_t         recvfrom   = (group_rank - 1 + group_size) % group_size;
+    int                step;
+    if (UCC_INPROGRESS == ucc_tl_ucp_test(task)) {
+        return task->super.super.status;
+    }
+
+    while (task->send_posted < group_size - 1) {
+        step = task->send_posted;
+        ucc_tl_ucp_send_nb((void*)((ptrdiff_t)rbuf +
+                           ((group_rank - step + group_size) % group_size)
```
Comment (manjugv) [inline src/components/tl/ucp/allgather/allgather_ring.c:36]: Is it required to send to self for correctness?

**Comment by manjugv** [comment]
@vspetrov Looks good. I have minor comments. Between, I'm assuming you ran data (MPI) integrity tests.

File: src/components/mc/cuda/mc_cuda.c (hunk 7 above)
Comment (bureddy) [reply src/components/mc/cuda/mc_cuda.c:114]: I think driver knows the address ranges for memory its allocated. cuda unified virtual addressing is the default on all platforms for a long time now. I think we can ignore that.

File: src/components/tl/ucp/allgather/allgather_ring.c (hunk 9 above)
Comment (vspetrov) [reply src/components/tl/ucp/allgather/allgather_ring.c:36]: dest buffer is different from source in general - in that case we need to send_self (we could use mc_memcpy, but adding another branch and reducing code readability - ucp will do memcpy underhoods anyways).

**Review by manjugv — APPROVED** [review]

**Comment by manjugv** [comment]
@vspetrov Can you please rebase?

File: src/components/mc/cuda/mc_cuda.c (hunk 7 above)
Comment (vspetrov) [reply src/components/mc/cuda/mc_cuda.c:114]: ok, changed to cudaMemcpyDefault

**Review by bureddy — APPROVED** [review]

**Pull request merged**